lib/pid/include/PID.h

```diff
@@ -1,282 +1,43 @@
-  float getPParam();
-  float getIParam();
-  float getDParam();
-  float getATunePParam();
-  float getATuneIParam();
-  float getATuneDParam();
-  float getSetPoint();
+  PID(uint32_t proportionalGain, uint32_t intregralGain, uint32_t derivativeGain, int32_t lowerBound,
+      int32_t upperBound, float deadzone, bool useAntiDerivativeKickback = true);
+
+  // WARNING: THESE ARE CALLED IN THE CANRX THREAD
+  mbed_error_status_t updateProportionalGain(const uint32_t p);
+  mbed_error_status_t updateIntegralGain(const uint32_t i);
+  mbed_error_status_t updateDerivativeGain(const uint32_t d);
+  mbed_error_status_t updateDeadzone(const float deadzone);
+
+  // WARNING: THESE ARE CALLED IN CANRX THREAD
+  uint32_t reportProportionalGain() const;
+  uint32_t reportIntegralGain() const;
+  uint32_t reportDerivativeGain() const;
+  float reportDeadzone() const;
+
+  // WARNING: THESE ARE CALLED IN CANRX THREAD
+  void reset();
+  // WARNING: COMPUTE IS CALLED IN MAIN THREAD
+  float compute(float setPoint, float processVariable);
 
  private:
-  bool usingFeedForward;
-  bool inAuto;
-
-  // Actual tuning parameters used in PID calculation.
-  float Kc_;
-  float tauR_;
-  float tauD_;
-
-  // Tuning parameters calculated from autotuning function
-  float autoTuneKc_;
-  float autoTuneTauR_;
-  float autoTuneTauD_;
-
-  // Input/Output pointers for AutoTuning
-  void *output_;
-  float *input_;
-
-  // Store type of actuator PID is being used on
-  int actuatorType_;
-
-  // Struct with all the autotune parameters
-  t_AutoTuneConfig AutoTuneConfig;
-
-  // Raw tuning parameters.
-  float pParam_;
-  float iParam_;
-  float dParam_;
-
-  // The point we want to reach.
-  float setPoint_;
-  // The thing we measure.
-  float processVariable_;
-  float prevProcessVariable_;
-  // The output that affects the process variable.
-  float controllerOutput_;
-  float prevControllerOutput_;
-
-  // We work in % for calculations so these will scale from
-  // real world values to 0-100% and back again.
-  float inMin_;
-  float inMax_;
-  float inSpan_;
-  float outMin_;
-  float outMax_;
-  float outSpan_;
-
-  // The accumulated error, i.e. integral.
-  float accError_;
-  // The allowed error range for error to be rounded to 0.0
-  float deadZoneError_;
-  // The controller output bias.
-  float bias_;
-
-  // The interval between samples.
-  std::chrono::duration<float> tSample_;
-
-  // Controller output as a real world value.
-  volatile float realOutput_;
+  // Watch out for deadlock if editing this code
+  Mutex m_timerMutex;
+  // TIMER SHOULD BE ATOMIC/MUTEX PROTECTED
+  Timer m_timer;
+  std::atomic<uint32_t> m_PGain, m_IGain, m_DGain;
+  const int32_t m_lowerBound, m_upperBound;
+  // NO ATOMIC INCREMENT FOR FLOATS?????????/
+  std::atomic<float> m_deadzone;
+  mutable std::atomic<float> m_IPath;
+  std::atomic<float> m_pastError, m_pastPV;
+  const bool m_antiKickback;
+  float computePPath(float error) const;
```

Comment:
Just for my learning but what does "Path" mean her? Does it just mean it is calculating the new P value?

Reply:
yeah if by P value you mean P_Gain * error